cli/analyze/analyze.go

```diff
@@ -0,0 +1,559 @@
+	visited := map[string]bool{}
+	for len(frontier) > 0 {
+		name := frontier[len(frontier)-1]
+		frontier = frontier[:len(frontier)-1]
+		if visited[name] {
+			continue
+		}
+		visited[name] = true
+		for _, r := range g.DirectDependents[name] {
+			frontier = append(frontier, *r.Name)
+		}
+	}
+
+	return len(visited)
+}
+
+// LongestPath returns the longest dependency path from one node in the graph to
+// any other node.
+func (g *DependencyGraph) LongestPath() []string {
+	longestPathLength := 0
+	var longestPathEnd *string
+
+	length := map[string]int{}
+	pred := map[string]*string{}
+	e := g.EdgeSet()
+	for _, m := range g.TopologicalSort() {
+		for n := range e.Incoming[m] {
+			l := length[n] + 1
+			if l > length[m] {
+				length[m] = l
+				n := n
+				pred[m] = &n
+			}
+			if l > longestPathLength {
+				longestPathLength = l
+				m := m
+				longestPathEnd = &m
+			}
+		}
+	}
+	if longestPathEnd == nil {
+		return nil
+	}
+	path := []string{*longestPathEnd}
+	for {
+		p := pred[path[len(path)-1]]
+		if p == nil {
+			break
+		}
+		path = append(path, *p)
+	}
+	reverseSlice(path)
+	return path
+}
+
+// Roots returns all labels in the graph which have no direct dependents.
+func (g *DependencyGraph) Roots() []string {
+	var roots []string
+	for name := range g.Rules {
+		if len(g.DirectDependents[name]) == 0 {
+			roots = append(roots, name)
+		}
+	}
+	return roots
+}
+
+func (g *DependencyGraph) EdgeSet() *EdgeSet {
+	outgoing := map[string]map[string]bool{}
+	incoming := map[string]map[string]bool{}
+	s := g.Roots()
+	for len(s) > 0 {
+		n := s[len(s)-1]
+		s = s[:len(s)-1]
+		if _, ok := outgoing[n]; ok {
+			continue // already visited
+		}
+		rule := g.Rules[n]
+		if rule == nil {
+			continue
+		}
+		for _, m := range rule.RuleInput {
+			if outgoing[n] == nil {
+				outgoing[n] = map[string]bool{}
+			}
+			if incoming[m] == nil {
+				incoming[m] = map[string]bool{}
+			}
+			outgoing[n][m] = true
+			incoming[m][n] = true
+
+			s = append(s, m)
+		}
+	}
+	return &EdgeSet{Outgoing: outgoing, Incoming: incoming}
+}
+
+// TopologicalSort returns all node labels in the graph sorted so that each node
+// is sorted before any of its transitive deps.
+func (g *DependencyGraph) TopologicalSort() []string {
+	// https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm
+	var out []string
+	s := g.Roots()
+	e := g.EdgeSet()
+	for len(s) > 0 {
+		n := s[len(s)-1]
+		s = s[:len(s)-1]
+		out = append(out, n)
+		for m := anyKey(e.Outgoing[n]); m != nil; m = anyKey(e.Outgoing[n]) {
+			e.Remove(n, *m)
+			if len(e.Incoming[*m]) == 0 {
+				s = append(s, *m)
+			}
+		}
+	}
+	return out
+}
+
+func getAllSourceFilePaths(dir string) (map[string]bool, error) {
+	lines, err := getOutputLines(dir, "git", "ls-files", "--exclude-standard")
+	if err != nil {
+		return nil, fmt.Errorf("git ls-files failed: %s", err)
+	}
+	return makeSet(lines), nil
+}
+
+func getOutputLines(dir, executable string, args ...string) ([]string, error) {
```

Comment:
This is pretty darn close to https://pkg.go.dev/os/exec#Cmd.Output .

Reply:
two differences -
- It properly splits output lines, since the most obvious implementation is `strings.Split(string(output), "\n")`, but that's incorrect since it results in the last line being blank (due to a trailing newline)
- The `err` returned by `cmd.Output()` doesn't include the command's stderr, which is important for understanding what went wrong. Instead the error message is something like "exit code 1". This func instead returns the command's stderr as the error, which is suitable for these git commands since they fail fast with succinct one-line error messages.